net/httpclient.go

```diff
@@ -0,0 +1,234 @@
+}
+
+// NewTransport creates a wrapped http.Transport, with regular DNS
+// lookups using CloseIdleConnections on every IdleConnTimeout. You
+// can optionally add tracing. On teardown you have to close the quit
+// channel or you will leak a goroutine.
+func NewTransport(options Options, quit <-chan struct{}) *Transport {
+	// set default tracer
+	if options.Tracer == nil {
+		options.Tracer = &opentracing.NoopTracer{}
+	}
+
+	// set timeout defaults
+	if options.TLSHandshakeTimeout == 0 {
+		options.TLSHandshakeTimeout = options.Timeout
+	}
+	if options.IdleConnTimeout == 0 {
+		options.IdleConnTimeout = options.Timeout
+	}
+	if options.ResponseHeaderTimeout == 0 {
+		options.ResponseHeaderTimeout = options.Timeout
+	}
+	if options.ExpectContinueTimeout == 0 {
+		options.ExpectContinueTimeout = options.Timeout
+	}
+
+	htransport := &http.Transport{
+		DisableKeepAlives:      options.DisableKeepAlives,
+		DisableCompression:     options.DisableCompression,
+		ForceAttemptHTTP2:      options.ForceAttemptHTTP2,
+		MaxIdleConns:           options.MaxIdleConns,
+		MaxIdleConnsPerHost:    options.MaxIdleConnsPerHost,
+		MaxConnsPerHost:        options.MaxConnsPerHost,
+		WriteBufferSize:        options.WriteBufferSize,
+		ReadBufferSize:         options.ReadBufferSize,
+		MaxResponseHeaderBytes: options.MaxResponseHeaderBytes,
+		ResponseHeaderTimeout:  options.ResponseHeaderTimeout,
+		TLSHandshakeTimeout:    options.TLSHandshakeTimeout,
+		IdleConnTimeout:        options.IdleConnTimeout,
+		ExpectContinueTimeout:  options.ExpectContinueTimeout,
+	}
+
+	go func() {
+		for {
+			select {
+			case <-time.After(options.IdleConnTimeout):
+				htransport.CloseIdleConnections()
+			case <-quit:
+				return
+			}
+		}
+	}()
+
+	return &Transport{
+		tr:     htransport,
+		tracer: options.Tracer,
+	}
+}
+
+// WithSpanName sets the name of the span, if you have an enabled
+// tracing Transport.
+func WithSpanName(t *Transport, spanName string) *Transport {
+	tt := t.shallowCopy()
+	tt.spanName = spanName
+	return tt
+}
+
+// WithComponentTag sets the component name, if you have an enabled
+// tracing Transport.
+func WithComponentTag(t *Transport, componentName string) *Transport {
+	tt := t.shallowCopy()
+	tt.componentName = componentName
+	return tt
+}
+
+// WithBearerToken adds an Authorization header with "Bearer " prefix
+// and add the given bearerToken as value to all requests. To regular
+// update your token you need to call this method and use the returned
+// Transport.
+func WithBearerToken(t *Transport, bearerToken string) *Transport {
+	tt := t.shallowCopy()
+	tt.bearerToken = bearerToken
+	return tt
+}
+
+func (t *Transport) shallowCopy() *Transport {
+	tt := *t
+	return &tt
+}
+
+// Do the request with tracing, bearer token injection and add client
+// tracing: DNS, TCP/IP, TLS handshake, connection pool access. Client
+// traces are added as logs into the created span.
+func (t *Transport) Do(req *http.Request) (*http.Response, error) {
```

Comment:
since this is more similar in behavior to the net/http.RoundTripper.RoundTrip(), let's name this function RoundTrip, and in a next pass we can wrap it then with a client that behaves like net/http.Client.Do(), e.g. following redirects